binary/build-binary.py

```diff
@@ -0,0 +1,256 @@
+# Copyright (C) 2020 Simon Biggs
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+import subprocess
+import shutil
+import sys
+import urllib.request
+import zipfile
+import functools
+
+import tomlkit
+
+HERE = pathlib.Path("__file__").parent.resolve()
+REPO_ROOT = HERE.parent
+PYPROJECT_TOML_PATH = REPO_ROOT.joinpath("pyproject.toml")
+LICENSE_PATH = REPO_ROOT.joinpath("LICENSE")
+
+DIST = REPO_ROOT.joinpath("dist")
+
+BUILD = HERE.joinpath("build")
+WHEELS = BUILD.joinpath("wheels")
+
+DOWNLOADS = BUILD.joinpath("downloads")
+PYTHON_EMBED_URL = (
+    "https://www.python.org/ftp/python/3.7.9/python-3.7.9-embed-amd64.zip"
+)
+PYTHON_EMBED_PATH = DOWNLOADS.joinpath("python-embed.zip")
+GET_PIP_URL = "https://bootstrap.pypa.io/get-pip.py"
+GET_PIP_PATH = DOWNLOADS.joinpath("get-pip.py")
+
+BUILD_PYTHON_EMBED = BUILD.joinpath("python-embed")
+BUILD_PYTHON_EMBED_XZTAR = BUILD_PYTHON_EMBED.with_suffix(".tar.xz")
+
+BUILD_DIST = BUILD.joinpath("dist")
+
+
+def main():
+    """Builds ``PyMedPhysGUI-vX.Y.Z.exe`` that when run extracts an
+    embedded Python distro, spins up the PyMedPhys GUI, and lastly
+    exposes the CLI via a ``pymedphys.bat``.
+
+    Note
+    ----
+    Under the hood this does use pyinstaller to create the exe. However
+    the pyinstaller section of the resulting code just ends up being an
+    "installer script". PyMedPhys is not installed or utilised within
+    the pyinstaller script itself. Instead, PyMedPhys is installed
+    within the embedded Python distribution which is bundled as a
+    compressed file within the executable. See
+    ``pyinstaller-bundle-script.py`` for more details on what occurs
+    when the executable is run.
+
+    """
+    prepend, append, one_file_mode = _linux_and_windows_support()
+
+    _build_and_collate_wheels(prepend)
+    _download_and_extract_embedded_python()
+    _get_pip(prepend)
+    _install_pymedphys_in_offline_mode(prepend)
+    _create_compressed_python_embed()
+
+    _run_pyinstaller_to_build_the_exe(prepend, append, one_file_mode)
+
+
+def _linux_and_windows_support():
+    """Shimming to allow for building the exe on Linux with wine as well as on Windows.
+
+    If the OS isn't Windows all python calls a prepended with ``wine``.
+    Also, if it is being built for Linux --onefile mode is not used due
```

Comment:
the CI binary task is excluded for linux and MacOS.
Is the implication here that there is partial progress towards building a linux binary?
Or does "being built for Linux" mean: being built on Linux for MS Windows
Or "being built for Linux assuming that Wine is available on the linux host"
Is this going to work for MacOS?  Apparently Wine is available for MacOS, but MacOS isn't Linux.
It is a *nix, but it isn't Linux.

Reply:
> Or does "being built for Linux" mean: being built on Linux for MS Windows

That one :).

This executable building is to build a Windows exe, so it's the only CI that should be doing it. Many things in this script are Windows only. At the end of the day, the "wine" is just so that I could "cross-compile". Don't have Windows at home, so wanted to be able to test that it was going to work in the CI without having to wait for the CI.

It's true that MacOS may work. And yes, the should be "built for testing on Linux"

```suggestion
    Also, if it is being built for testing on Linux --onefile mode is not used due
```